browser/search_engines/search_engine_provider_service.cc

```diff
@@ -12,52 +15,64 @@
+    : profile_(profile),
+      template_url_service_(
+          TemplateURLServiceFactory::GetForProfile(profile_)) {
   use_alternative_search_engine_provider_.Init(
       kUseAlternativeSearchEngineProvider,
-      otr_profile_->GetOriginalProfile()->GetPrefs(),
+      profile_->GetOriginalProfile()->GetPrefs(),
       base::Bind(&SearchEngineProviderService::OnPreferenceChanged,
                  base::Unretained(this)));
 
-  std::vector<TemplateURLPrepopulateData::BravePrepopulatedEngineID>
-      alt_search_providers = {
-          TemplateURLPrepopulateData::PREPOPULATED_ENGINE_ID_DUCKDUCKGO,
-          TemplateURLPrepopulateData::PREPOPULATED_ENGINE_ID_DUCKDUCKGO_DE,
-          TemplateURLPrepopulateData::
-              PREPOPULATED_ENGINE_ID_DUCKDUCKGO_AU_NZ_IE
-      };
+  std::unique_ptr<TemplateURLData> data =
+      brave::GetDDGTemplateURLData(profile->GetPrefs());
+  alternative_search_engine_url_.reset(new TemplateURL(*data));
+  observation_.Observe(template_url_service_);
+}
+
+SearchEngineProviderService::~SearchEngineProviderService() = default;
+
+void SearchEngineProviderService::OnTemplateURLServiceChanged() {
+  // When this change is came from pref changing, we don't need to control
+  // prefs again.
+  if (ignore_template_url_service_changing_)
+    return;
 
-  std::unique_ptr<TemplateURLData> data;
-  for (const auto& id : alt_search_providers) {
-    data = TemplateURLPrepopulateData::GetPrepopulatedEngine(
-        otr_profile->GetPrefs(), id);
-    if (data)
+  // The purpose of below code is toggling alternative prefs
+  // when user changes from ddg to different search engine provider
+  // (or vice versa) from settings ui.
+  bool is_ddg_is_set = false;
+  switch (template_url_service_->GetDefaultSearchProvider()
+              ->data()
+              .prepopulate_id) {
+    case TemplateURLPrepopulateData::PREPOPULATED_ENGINE_ID_DUCKDUCKGO:
+    case TemplateURLPrepopulateData::PREPOPULATED_ENGINE_ID_DUCKDUCKGO_DE:
+    case TemplateURLPrepopulateData::PREPOPULATED_ENGINE_ID_DUCKDUCKGO_AU_NZ_IE:
+      is_ddg_is_set = true;
+      break;
+    default:
       break;
   }
 
-  // There should ALWAYS be one entry
-  DCHECK(data);
-  alternative_search_engine_url_.reset(new TemplateURL(*data));
-}
-
-SearchEngineProviderService::~SearchEngineProviderService() {
+  if (UseAlternativeSearchEngineProvider() || is_ddg_is_set)
+    brave::ToggleUseAlternativeSearchEngineProvider(profile_);
 }
 
 void SearchEngineProviderService::OnPreferenceChanged(
     const std::string& pref_name) {
   DCHECK(pref_name == kUseAlternativeSearchEngineProvider);
-  DCHECK(!brave::IsRegionForQwant(otr_profile_));
 
-  OnUseAlternativeSearchEngineProviderChanged();
+  // When this call is from setting's change, we don't need to set provider
+  // again.
+  if (ignore_template_url_service_changing_)
+    return;
+
+  base::AutoReset<bool> reset(&ignore_template_url_service_changing_, true);
```

Comment:
This is a super nice base library offering 🙂 Had never worked with this before! Really nice to temporarily change the value and then it resets after going out of scope